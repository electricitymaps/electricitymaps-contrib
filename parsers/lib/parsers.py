import importlib

from electricitymap.contrib.config import EXCHANGES_CONFIG, ZONES_CONFIG

# Prepare all parsers
CONSUMPTION_PARSERS = {}
PRODUCTION_PARSERS = {}
PRODUCTION_PER_MODE_FORECAST_PARSERS = {}
PRODUCTION_PER_UNIT_PARSERS = {}
EXCHANGE_PARSERS = {}
PRICE_PARSERS = {}
CONSUMPTION_FORECAST_PARSERS = {}
GENERATION_FORECAST_PARSERS = {}
EXCHANGE_FORECAST_PARSERS = {}
PRODUCTION_CAPACITY_PARSERS = {}
DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}
REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS = {}

PARSER_KEY_TO_DICT = {
    "consumption": CONSUMPTION_PARSERS,
    "production": PRODUCTION_PARSERS,
    "productionPerUnit": PRODUCTION_PER_UNIT_PARSERS,
    "productionPerModeForecast": PRODUCTION_PER_MODE_FORECAST_PARSERS,
    "exchange": EXCHANGE_PARSERS,
    "price": PRICE_PARSERS,
    "dayaheadLocationalMarginalPrice": DAYAHEAD_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "realtimeLocationalMarginalPrice": REALTIME_LOCATIONAL_MARGINAL_PRICE_PARSERS,
    "consumptionForecast": CONSUMPTION_FORECAST_PARSERS,
    "generationForecast": GENERATION_FORECAST_PARSERS,
    "exchangeForecast": EXCHANGE_FORECAST_PARSERS,
    "productionCapacity": PRODUCTION_CAPACITY_PARSERS,
}


def _parser_key_to_parser_folder(parser_key: str):
    return (
        "electricitymap.contrib.capacity_parsers"
        if parser_key == "productionCapacity"
        else "parsers"
    )


# Read all zones
for zone_id, zone_config in ZONES_CONFIG.items():
    for parser_key, v in zone_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(
            f"{_parser_key_to_parser_folder(parser_key)}.{mod_name}"
        )
        PARSER_KEY_TO_DICT[parser_key][zone_id] = getattr(mod, fun_name)


# Read all exchanges
for exchange_id, exchange_config in EXCHANGES_CONFIG.items():
    for parser_key, v in exchange_config.get("parsers", {}).items():
        mod_name, fun_name = v.split(".")
        mod = importlib.import_module(f"parsers.{mod_name}")
        PARSER_KEY_TO_DICT[parser_key][exchange_id] = getattr(mod, fun_name)
