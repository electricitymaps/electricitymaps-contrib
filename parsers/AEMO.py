import io
import re
import zipfile
from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from urllib.parse import urljoin
from zoneinfo import ZoneInfo

import pandas as pd
from bs4 import BeautifulSoup
from requests import Session

from electricitymap.contrib.lib.models.event_lists import TotalConsumptionList
from electricitymap.contrib.lib.models.events import EventSourceType
from electricitymap.contrib.lib.types import ZoneKey

SOURCE = "aemo.com.au"

ZONE_KEY_TO_REGION = {
    "AU-NSW": "NSW1",
    "AU-QLD": "QLD1",
    "AU-SA": "SA1",
    "AU-TAS": "TAS1",
    "AU-VIC": "VIC1",
    "AU-WA": "WEM",  # This zone is not implemented yet
}

ZONE_KEY_TO_TIMEZONE = {
    "AU-NSW": ZoneInfo("Australia/Sydney"),
    "AU-QLD": ZoneInfo("Australia/Brisbane"),
    "AU-SA": ZoneInfo("Australia/Adelaide"),
    "AU-TAS": ZoneInfo("Australia/Hobart"),
    "AU-VIC": ZoneInfo("Australia/Melbourne"),
    "AU-WA": ZoneInfo("Australia/Perth"),  # This zone is not implemented yet
}

# TODO, what about the other zone in Australia (AU-WA)? Check remaining zone


def find_document(session, target_datetime):
    # Fetch the directory listing
    base_url = "http://nemweb.com.au/Reports/CURRENT/Operational_Demand/FORECAST_HH/"
    response = session.get(base_url)

    # Parse with BeautifulSoup
    soup = BeautifulSoup(response.text, "html.parser")

    # String datetime
    target_date_str = target_datetime.strftime("%Y%m%d")

    # Find matching links
    matching_links = soup.find_all(
        "a",
        href=re.compile(
            rf"PUBLIC_FORECAST_OPERATIONAL_DEMAND_HH_{target_date_str}\d+_\d+\.zip"
        ),
    )

    if matching_links:
        target_link = matching_links[-1]  # Get the last (most recent) link
        full_url = urljoin(base_url, target_link["href"])  # Construct full URL

        file_response = session.get(full_url)

        with zipfile.ZipFile(io.BytesIO(file_response.content)) as z:
            csv_filename = z.namelist()[0]
            with z.open(csv_filename) as f:
                df = pd.read_csv(f)
                return df
    else:
        print("No matching files found")


def fetch_consumption_forecast(
    zone_key: ZoneKey,  # "AU-NSW", "AU-QLD", "AU-SA", "AU-TAS", "AU-VIC", "AU-WA"
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    """Consumption forecast in MW every half an hour for 10 days ahead.
    Only for NSW1, QND1, SA1, TAS1, VIC1 zones."""
    session = session or Session()

    if target_datetime is None:
        target_datetime = datetime.now(tz=ZONE_KEY_TO_TIMEZONE[zone_key])

    df = find_document(session, target_datetime)

    # Transform dataframe
    df.columns = df.iloc[0]
    df = df.iloc[1:-1].reset_index(drop=True)

    #
    region = ZONE_KEY_TO_REGION.get(zone_key)
    all_consumption_events = df[
        df["REGIONID"] == region
    ]  # all events with a datetime and a consumption value
    consumption_list = TotalConsumptionList(logger)
    for _, event in all_consumption_events.iterrows():
        datetime_object = datetime.strptime(
            event["INTERVAL_DATETIME"], "%Y/%m/%d %H:%M:%S"
        ).replace(tzinfo=ZONE_KEY_TO_TIMEZONE[zone_key])

        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime_object,
            consumption=float(
                event["OPERATIONAL_DEMAND_POE50"]
            ),  # 50% probability of exceedance operational demand forecast value
            source=SOURCE,
            sourceType=EventSourceType.forecasted,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the electricityMap backend, but handy for testing."""

    print(fetch_consumption_forecast("AU-NSW"))
    print(fetch_consumption_forecast("AU-QLD"))
    print(fetch_consumption_forecast("AU-SA"))
    print(fetch_consumption_forecast("AU-TAS"))
    print(fetch_consumption_forecast("AU-VIC"))
    print(
        fetch_consumption_forecast("AU-WA")
    )  # Not implemented yet. It returns an empty list
