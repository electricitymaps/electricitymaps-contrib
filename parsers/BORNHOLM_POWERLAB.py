#!/usr/bin/env python3
from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

PRODUCTION_MAPPING = {
    "wind": "wind_turbines",
    "biomass": "factory",
    "solar": "solar_cells",
}
LATEST_DATA_URL = "https://bornholm.powerlab.dk/visualizer/latestdata"

SOURCE = "bornholm.powerlab.dk"
TIMEZONE = ZoneInfo("UTC")


def fetch_production(
    zone_key: ZoneKey = ZoneKey("DK-BHM"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    production = ProductionBreakdownList(logger)
    production.append(
        zoneKey=zone_key,
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        production=ProductionMix(
            biomass=response["sub"]["factory"],
            solar=response["sub"]["solar_cells"],
            wind=response["sub"]["wind_turbines"],
        ),
    )
    return production.to_list()


def fetch_exchange(
    zone_key1: ZoneKey = ZoneKey("DK-BHM"),
    zone_key2: ZoneKey = ZoneKey("SE-SE4"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known power exchange (in MW) between two countries."""
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    if session is None:
        session = Session()
    response = session.get(LATEST_DATA_URL).json()
    exchange = ExchangeList(logger)
    sorted_zone_keys = sorted([zone_key1, zone_key2])
    direction = 1 if zone_key1 == sorted_zone_keys[0] else -1
    exchange.append(
        zoneKey=ZoneKey("->".join(sorted_zone_keys)),
        datetime=datetime.fromtimestamp(response["latest"], tz=TIMEZONE),
        source=SOURCE,
        netFlow=direction * response["sub"]["seacable"],
    )

    return exchange.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
    print("fetch_exchange(DK-BHM, SE-SE4) ->")
    print(fetch_exchange())
