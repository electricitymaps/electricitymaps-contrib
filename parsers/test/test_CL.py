from datetime import datetime, timezone
from importlib import resources
from json import loads

import pytest
from requests_mock import GET

from electricitymap.contrib.lib.types import ZoneKey
from parsers.CL import API_BASE_URL, fetch_production


@pytest.fixture(autouse=True)
def mock_response(adapter):
    url = f"{API_BASE_URL}fecha__gte=2024-02-23&fecha__lte=2024-02-24"
    adapter.register_uri(
        GET,
        url,
        json=loads(
            resources.files("parsers.test.mocks.CL")
            .joinpath("response_historical_20240224.json")
            .read_text()
        ),
    )


def test_snapshot_historical_data(session, snapshot):
    target_datetime = datetime(2024, 2, 24, 0, 0, 0, tzinfo=timezone.utc)

    assert snapshot == fetch_production(
        ZoneKey("CL-SEN"),
        session=session,
        target_datetime=target_datetime,
    )
