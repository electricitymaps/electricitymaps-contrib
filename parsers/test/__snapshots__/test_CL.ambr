# serializer version: 1
# name: test_snapshot_historical_data
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 53.14,
        'hydro': 2700.906657,
        'solar': 100.000032,
        'unknown': 5268.170499,
        'wind': 1280.780885,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 53.26,
        'hydro': 2462.475625,
        'solar': 81.105925,
        'unknown': 5170.743269,
        'wind': 1189.91521,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 53.79,
        'hydro': 2379.239039,
        'solar': 8.300033,
        'unknown': 5117.478787,
        'wind': 1117.397269,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 3, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 54.38,
        'hydro': 2354.243241,
        'solar': 59.902852,
        'unknown': 5094.783766,
        'wind': 1090.907371,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 4, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 53.56,
        'hydro': 2283.744427,
        'solar': 120.724776,
        'unknown': 5087.732329,
        'wind': 1084.973849,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 5, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 53.56,
        'hydro': 2303.131467,
        'solar': 122.795107,
        'unknown': 5009.660575,
        'wind': 1155.499893,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 6, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 54.17,
        'hydro': 2410.308194,
        'solar': 125.61644,
        'unknown': 5045.755455,
        'wind': 1224.985927,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 7, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 54.96,
        'hydro': 2631.740156,
        'solar': 245.078277,
        'unknown': 4924.854536,
        'wind': 1302.988293,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 8, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 54.56,
        'hydro': 2611.38258,
        'solar': 2278.50336,
        'unknown': 3316.625734,
        'wind': 1288.618092,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 9, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 52.9,
        'hydro': 2576.73036,
        'solar': 4210.016137,
        'unknown': 2046.03737,
        'wind': 1075.449985,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 10, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 52.15,
        'hydro': 2319.68927,
        'solar': 5132.126592,
        'unknown': 1755.337808,
        'wind': 974.676364,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 11, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 49.11,
        'hydro': 2234.245319,
        'solar': 5484.558118,
        'unknown': 1696.242462,
        'wind': 932.930836,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 48.82,
        'hydro': 2172.62459,
        'solar': 5671.553406,
        'unknown': 1678.230359,
        'wind': 931.12142,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 13, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 49.19,
        'hydro': 2144.243675,
        'solar': 5578.896809,
        'unknown': 1736.036832,
        'wind': 1099.386722,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 14, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 46.03,
        'hydro': 2263.655016,
        'solar': 5422.101832,
        'unknown': 1778.553306,
        'wind': 1230.231538,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 15, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 42.12,
        'hydro': 2246.264277,
        'solar': 5389.586746,
        'unknown': 1751.539326,
        'wind': 1361.218869,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 16, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 42.29,
        'hydro': 2129.363594,
        'solar': 5202.744165,
        'unknown': 1878.335343,
        'wind': 1421.011156,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 17, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 43.37,
        'hydro': 2269.306443,
        'solar': 4613.716848,
        'unknown': 2089.748741,
        'wind': 1461.531899,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 18, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 42.92,
        'hydro': 2130.077228,
        'solar': 4030.796152,
        'unknown': 2880.643438,
        'wind': 1285.360625,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 19, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 49.96,
        'hydro': 2679.453131,
        'solar': 1850.659698,
        'unknown': 4302.627242,
        'wind': 1485.318669,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 20, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 52.22,
        'hydro': 3204.530755,
        'solar': 238.576226,
        'unknown': 5535.749833,
        'wind': 1607.077091,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 21, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 53.13,
        'hydro': 3339.080773,
        'solar': 183.033445,
        'unknown': 5590.121261,
        'wind': 1596.874509,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 22, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 53.31,
        'hydro': 3397.70348,
        'solar': 126.82,
        'unknown': 5451.267958,
        'wind': 1475.910419,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 23, 23, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 54.39,
        'hydro': 3050.825042,
        'solar': 124.8,
        'unknown': 5427.384926,
        'wind': 1468.875217,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 55.44,
        'hydro': 2921.135053,
        'solar': 97.9,
        'unknown': 5242.05276,
        'wind': 1394.006656,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 54.72,
        'hydro': 2873.639079,
        'solar': 12.0,
        'unknown': 5092.276621,
        'wind': 1278.197109,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 45.19,
        'hydro': 2779.006138,
        'solar': 3.3e-05,
        'unknown': 5017.079483,
        'wind': 1255.251476,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 3, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 36.36,
        'hydro': 2535.719159,
        'solar': 0.002682,
        'unknown': 5161.346538,
        'wind': 1216.978644,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 4, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 50.56,
        'hydro': 2455.683144,
        'solar': 0.666425,
        'unknown': 5156.500387,
        'wind': 1190.1354,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 5, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 50.6,
        'hydro': 2344.02243,
        'solar': 2.78778,
        'unknown': 5173.181655,
        'wind': 1265.949049,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 6, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 49.93,
        'hydro': 2436.928362,
        'solar': 4.671609,
        'unknown': 5066.064531,
        'wind': 1340.589791,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 7, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 50.17,
        'hydro': 2612.45783,
        'solar': 128.868019,
        'unknown': 4821.509308,
        'wind': 1252.866856,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 8, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 49.77,
        'hydro': 2416.600953,
        'solar': 2229.983225,
        'unknown': 3304.393787,
        'wind': 980.684729,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 9, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 48.89,
        'hydro': 2422.68662,
        'solar': 4072.759647,
        'unknown': 2101.012526,
        'wind': 738.109111,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 10, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 50.32,
        'hydro': 2125.226312,
        'solar': 4940.913316,
        'unknown': 1942.611623,
        'wind': 586.726374,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 11, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 47.35,
        'hydro': 2150.054145,
        'solar': 5228.110012,
        'unknown': 1821.132904,
        'wind': 572.901276,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 45.0,
        'hydro': 2085.775046,
        'solar': 5508.17195,
        'unknown': 1796.212359,
        'wind': 473.008079,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 13, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 43.41,
        'hydro': 2192.233165,
        'solar': 5395.869378,
        'unknown': 1908.469311,
        'wind': 472.354431,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 2, 24, 14, 0, tzinfo=zoneinfo.ZoneInfo(key='Chile/Continental')),
      'production': dict({
        'geothermal': 43.09,
        'hydro': 2270.685839,
        'solar': 5182.93405,
        'unknown': 1935.465852,
        'wind': 649.912391,
      }),
      'source': 'coordinador.cl',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'CL-SEN',
    }),
  ])
# ---
